src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/changelist/OverseasEntityChangeService.java

```diff
@@ -0,0 +1,173 @@
+import uk.gov.companieshouse.api.model.update.OverseasEntityDataApi;
+import uk.gov.companieshouse.overseasentitiesapi.model.dto.AddressDto;
+import uk.gov.companieshouse.overseasentitiesapi.model.dto.EntityDto;
+import uk.gov.companieshouse.overseasentitiesapi.model.dto.EntityNameDto;
+import uk.gov.companieshouse.overseasentitiesapi.model.dto.OverseasEntitySubmissionDto;
+import uk.gov.companieshouse.overseasentitiesapi.model.updatesubmission.changelist.changes.*;
+import uk.gov.companieshouse.overseasentitiesapi.model.updatesubmission.changelist.commonmodels.CompanyIdentification;
+import uk.gov.companieshouse.overseasentitiesapi.validation.OverseasEntityChangeComparator;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class OverseasEntityChangeService {
+    private final OverseasEntityChangeComparator overseasEntityChangeComparator;
+
+    @Autowired
+    public OverseasEntityChangeService(OverseasEntityChangeComparator overseasEntityChangeComparator) {
+        this.overseasEntityChangeComparator = overseasEntityChangeComparator;
+    }
+
+    public List<Change> collateOverseasEntityChanges(
+            Pair<CompanyProfileApi, OverseasEntityDataApi> existingRegistration,
+            OverseasEntitySubmissionDto updateSubmission) {
+        List<Change> changes = new ArrayList<>();
+
+        if (existingRegistration != null && updateSubmission != null) {
+            addNonNullChangeToList(changes, retrieveEntityNameChange(existingRegistration, updateSubmission));
+            addNonNullChangeToList(changes, retrievePrincipalAddressChange(existingRegistration, updateSubmission));
+            addNonNullChangeToList(changes, retrieveCorrespondenceAddressChange(existingRegistration, updateSubmission));
+            addNonNullChangeToList(changes, retrieveCompanyIdentificationChange(existingRegistration, updateSubmission));
+            addNonNullChangeToList(changes, retrieveEmailAddressChange(existingRegistration, updateSubmission));
+        }
+
+        return changes;
+    }
+
+    private void addNonNullChangeToList(List<Change> changes, Change change) {
+        if (change != null) {
+            changes.add(change);
+        }
+    }
+
+    private EntityNameChange retrieveEntityNameChange(
+            Pair<CompanyProfileApi, OverseasEntityDataApi> existingRegistration,
+            OverseasEntitySubmissionDto updateSubmission) {
+        return overseasEntityChangeComparator.compareEntityName(
+                Optional.ofNullable(existingRegistration.getLeft())
+                        .map(CompanyProfileApi::getCompanyName)
+                        .orElse(null),
+                Optional.ofNullable(updateSubmission)
+                        .map(OverseasEntitySubmissionDto::getEntityName)
+                        .map(EntityNameDto::getName)
+                        .orElse(null));
+    }
+
+    private PrincipalAddressChange retrievePrincipalAddressChange(
```

Comment:
This logic is quite complex, I'm assuming we're trying to cover the following scenarios:

- If updateSubmission == null or there's no serviceAddress or principalAddress, we return null
- If the serviceAddress == principalAddress then we default to the serviceAddress
- if the serviceAddress != principalAddress then we default to the principalAddress

Why not default to the principalAddress?

Reply:
So the logic is meant to flow as follows:
1. There’s a flag in the updateSubmission that states if the service address is the same as the principal address (getServiceAddressSameAsPrincipalAddress)
2. If true: we want to use the service address in the comparison with the existing principal address, and construct a PrincipalAddressChange object based on that
3. If false: we use the principal address and follow the same process…
4. In any case where the getServiceAddressSameAsPrincipalAddress or the address is null, we still do the comparison but if there is no change with the existing then null is returned

Reply:
Thanks both for clarifying